### PR TITLE
Fix Google sign in silent takeover

### PR DESCRIPTION
**Pull request by smusoke**
**Context**: There's been [investigation](https://codedotorg.atlassian.net/browse/FND-1035?focusedCommentId=20998) on whether we can refactor our User account creation to have them migrated/multiauth by default. A result of that investigation was needing to address and fix the scenarios in which accounts were still currently failing to migrate. If this wasn't addressed and we moved forward with refactoring, those accounts would instead fail to be created in general.

**What**: This change has our `allows_section_takeover` function check **all** of the currently signing in OAuth `User`'s emails. Before the update we only looked at the `email`/`hashed_email` field on the `User` record, and the email fields on the `primary_contact_info`(which is the chronologically first `AuthenticationOption`). Now we check **all** of the associated `AuthenticationOption`'s emails.

**Why**: The intention is to fix Google OAuth's known sign in issue where a User calls silent takeover on itself. After we confirm a [`User` exists with the given **OAuth credentials**](https://github.com/code-dot-org/code-dot-org/blob/fnd-2029-update-allows-section-takeover-conditions/dashboard/app/controllers/omniauth_callbacks_controller.rb#L60) and passes `allows_section_takeover`, silent takeover is triggered. Silent takeover looks for another `User` by the **OAuth response email**. `allows_section_takeover` was previously failing to account for the other emails a `User` might have, and [the lookup](https://github.com/code-dot-org/code-dot-org/blob/fnd-2029-update-allows-section-takeover-conditions/dashboard/app/controllers/omniauth_callbacks_controller.rb#L382) was finding the exact same User. Silent takeover was then being ran on the same `User` it was being called by.

The result was the existing `User` record being soft deleted, and a new `AuthenticationOption` being created and associated to it. When an end-user tried to log in again, we would create a new `User` that would fail to "migrate" because there were conflicts with the new `AuthenticationOption`(duplicate `authentication_id`s) that is floating in limbo. 

## Links

- jira ticket: [FND-2029](https://codedotorg.atlassian.net/browse/FND-2029)

## Testing story

- Locally created a Clever User and associated a `google_oauth2` AuthenticationOption as well. Then logged in via Google to confirm that silent takeover wouldn't be triggered after the fix.
- Confirmed removing the fix triggers silent takeover.
- Locally ran the unit test creating the same scenario before and after the logic fix.

## Follow-up work

Continue to monitor if unmigrated `google_oauth2` Users are still be created after merge.

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: dashboard/app/controllers/omniauth_callbacks_controller.rb [hunk 1]
```diff
@@ -358,9 +358,13 @@ def allows_section_takeover(user)
     # OAuth providers do not necessarily provide student email addresses, so we
     # want to perform silent takeover on these accounts, but *only if* the
     # student hasn't made progress with the initial account
+    auth_options_emails = user.migrated? ?
+      user.authentication_options.select {|ao| !ao.hashed_email.blank? && ao.hashed_email} :
+      []
```
Comment (hacodeorg) [inline dashboard/app/controllers/omniauth_callbacks_controller.rb]: Could this be simplified as 
`has_auth_email = user.migrated? && user.authentication_options.any? {|ao| ao.hashed_email.present?}`

File: dashboard/test/controllers/omniauth_callbacks_controller_test.rb [hunk 2]
```diff
@@ -962,6 +962,26 @@ class OmniauthCallbacksControllerTest < ActionController::TestCase
...
+    @request.env['omniauth.auth'] = auth
+    @request.env['omniauth.params'] = {}
+
+    assert_does_not_destroy(User) do
+      get :google_oauth2
+    end
+    user.reload
+    assert_equal 'migrated', user.provider
+    auth_option_count = user.authentication_options.count
+    assert_equal 2, auth_option_count
```
Comment (hacodeorg) [inline dashboard/test/controllers/omniauth_callbacks_controller_test.rb:983]: Could you add a comment explaining what 2 authentication options we are expecting? Do we expect `get :google_oauth2` not creating a new authentication option?

File: dashboard/app/controllers/omniauth_callbacks_controller.rb (hunk 1 above)
Comment (smusoke) [reply dashboard/app/controllers/omniauth_callbacks_controller.rb]: `present?` is a good solution here. I'll update this.

**Review by hacodeorg — APPROVED** [review]
LGTM. Thank you.

**Pull request merged**